src/awkward/forms/unionform.py

```diff
@@ -220,8 +221,8 @@ def branch_depth(self):
 
     @property
     def fields(self):
-        fieldslists = [cont.fields for cont in self._contents]
-        return list(set.intersection(*[set(x) for x in fieldslists]))
+        field_counts = Counter([f for c in self._contents for f in c.fields])
+        return [f for f, n in field_counts.items() if n == len(self._contents)]
```

Comment:
The old algorithm was trying to say, "If a field name is in all contents, include it." The old algorithm may have lost field order, and that's a problem because this `fields` property should be ordered.

The new algorithm counts how many times a name is seen, and if that's equal to the number of union contents, it's a returned field. It definitely maintains field order, which is better than the old code. But is it always right? I suppose that one union content can't have the same field name twice, so if a field name ever reaches the number of union contents, it must have been in every content.

So I believe that it works! But an essential part of the proof is that a given field name can only appear in a union content zero or one times.

Reply:
Yes, this is an assumption. If we don't check this, we certainly should as it seems highly irregular (only L2 users could even access repeated fields, right?)